File: apps/desktop/src-tauri/src/crypto/hashing.rs
use sha2::{Sha256, Digest};

/// Compute SHA-256 hash, return as hex string.
pub fn sha256_hex(data: &[u8]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(data);
    let result = hasher.finalize();
    hex_encode(&result)
}

/// Compute incremental commitment: SHA-256(previous || nonce || data)
pub fn commitment_hash(previous: Option<&str>, nonce: &[u8], data: &[u8]) -> String {
    let mut hasher = Sha256::new();
    if let Some(prev) = previous {
        hasher.update(hex_decode(prev));
    }
    hasher.update(nonce);
    hasher.update(data);
    let result = hasher.finalize();
    hex_encode(&result)
}

/// Generate a random 256-bit nonce.
pub fn random_nonce() -> [u8; 32] {
    let mut nonce = [0u8; 32];
    getrandom(&mut nonce);
    nonce
}

fn getrandom(buf: &mut [u8]) {
    use rand::RngCore;
    rand::thread_rng().fill_bytes(buf);
}

fn hex_encode(bytes: &[u8]) -> String {
    bytes.iter().map(|b| format!("{:02x}", b)).collect()
}

fn hex_decode(hex: &str) -> Vec<u8> {
    (0..hex.len())
        .step_by(2)
        .map(|i| u8::from_str_radix(&hex[i..i + 2], 16).unwrap_or(0))
        .collect()
}

File: packages/core/src/crypto/hashing.ts
/**
 * Cryptographic hashing utilities for commitment chains.
 * Uses Web Crypto API — all functions are async.
 * Byte layout matches the Rust implementation exactly.
 */

export function hexEncode(bytes: Uint8Array): string {
  return Array.from(bytes)
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

export function hexDecode(hex: string): Uint8Array {
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < hex.length; i += 2) {
    bytes[i / 2] = parseInt(hex.substring(i, i + 2), 16);
  }
  return bytes;
}

function concatBytes(...arrays: Uint8Array[]): Uint8Array {
  const totalLength = arrays.reduce((sum, arr) => sum + arr.length, 0);
  const result = new Uint8Array(totalLength);
  let offset = 0;
  for (const arr of arrays) {
    result.set(arr, offset);
    offset += arr.length;
  }
  return result;
}

async function sha256(data: Uint8Array): Promise<ArrayBuffer> {
  // Copy into a fresh ArrayBuffer to satisfy TS 5.9 BufferSource constraints
  const buf = new ArrayBuffer(data.length);
  new Uint8Array(buf).set(data);
  return crypto.subtle.digest("SHA-256", buf);
}

/** Compute SHA-256 hash, return as hex string. */
export async function sha256Hex(
  data: string | Uint8Array,
): Promise<string> {
  const encoded =
    typeof data === "string" ? new TextEncoder().encode(data) : data;
  const hashBuffer = await sha256(encoded);
  return hexEncode(new Uint8Array(hashBuffer));
}

/** Compute incremental commitment: SHA-256(previous || nonce || data || documentHash?) */
export async function commitmentHash(
  previous: string | null,
  nonce: Uint8Array,
  data: Uint8Array,
  documentHash?: string,
): Promise<string> {
  const parts: Uint8Array[] = [];
  if (previous) {
    parts.push(hexDecode(previous));
  }
  parts.push(nonce);
  parts.push(data);
  if (documentHash) {
    parts.push(hexDecode(documentHash));
  }
  const combined = concatBytes(...parts);
  const hashBuffer = await sha256(combined);
  return hexEncode(new Uint8Array(hashBuffer));
}

/**
 * Compute content-binding commitment: SHA-256(chain_tip || "CONTENT_BINDING" || content_hash)
 * Binds the entire commitment chain to the final document content.
 */
export async function contentBindingHash(
  chainTip: string,
  contentHash: string,
): Promise<string> {
  const combined = concatBytes(
    hexDecode(chainTip),
    new TextEncoder().encode("CONTENT_BINDING"),
    hexDecode(contentHash),
  );
  const hashBuffer = await sha256(combined);
  return hexEncode(new Uint8Array(hashBuffer));
}
